### PR TITLE
Programmes-5755-RecEng

### PR DESCRIPTION
**Pull request by mattgreenham**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-5755

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/RecEng/RecEngService.php [hunk 1]
```diff
@@ -113,6 +123,7 @@ public function getRecommendations(
 
         $recEngRespose = $response->getBody()->getContents();
 
+
```
Comment (felipeparaujo) [inline src/RecEng/RecEngService.php]: WHITESPCAE!!!!!!!11111!!!!

**Comment by felipeparaujo** [comment]
looks alright, but as you renamed `getDuration()` to `showDuration()` you'll have to change that in the other amen objects that use the CTA (which is all of them)

**Comment by mattgreenham** [comment]
From conversation IRL, `getDuration()` hasn't been removed so no need to change

File: src/RecEng/RecEngService.php [hunk 2]
```diff
@@ -40,45 +42,54 @@ class RecEngService
...
      */
     public function getRecommendations(
         Programme $programme,
-        ?Episode $latestEpisode,
-        ?Episode $upcomingEpisode,
-        ?Episode $lastOnEpisode,
         int $limit = 2
     ): array {
-        $programmeEpisode = $this->getProgrammeEpisode($programme, $latestEpisode, $upcomingEpisode, $lastOnEpisode);
+        $programmeEpisode = $this->getProgrammeEpisode($programme);
```
Comment (mattgreenham) [inline src/RecEng/RecEngService.php]: Is it worth also caching this result or is the PPS caching sufficient?

File: app/Resources/js/lazyload.js [hunk 3]
```diff
@@ -126,6 +126,8 @@ define(['jquery-1.9'], function ($) {
                 success: function (data) {
                     var loaded_content = $($.parseHTML(data, document, true)).addClass(_this.options.lazy_css_state.loading);
                     content.replaceWith(loaded_content);
+                    loaded_content.addClass(_this.options.lazy_css_state.complete);
+                    loaded_content.removeClass(_this.options.lazy_css_state.loading);
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Should these be in the order, or should they be swapped?

File: app/Resources/views/recommendations/ds2013_recommends_footer.html.twig [hunk 4]
```diff
@@ -0,0 +1 @@
+<h2 class="footer__heading gel-double-pica-bold">{{- tr('you_might_also_like') -}}</h2>
```
Comment (votemike) [inline app/Resources/views/recommendations/ds2013_recommends_footer.html.twig]: Do 2013 pages not show recommendations?

File: src/Controller/RecommendationsController.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        $recommendations = $recEngService->getRecommendations($programme, 2);
+
+        if ($designSystem === 'ds_amen') {
+            return $this->renderWithChrome('recommendations/dsamen_recommends_footer.html.twig', [
+                'recommendations' => $recommendations,
+            ]);
+        }
+
+        if ($designSystem === 'ds_2013') {
+            return $this->renderWithChrome('ds2013_recommends_footer.html.twig', [
```
Comment (votemike) [inline src/Controller/RecommendationsController.php]: Does this not need to be prefixed with "recommendations/"?

File: tests/Controller/FindByPid/Tlec/TlecControllerClipsListTest.php [hunk 6]
```diff
@@ -20,8 +20,8 @@ public function setUp()
     public function testBrandWithNoClipsDoesntShowGrid()
     {
         $crawler = $this->client->request('GET', '/programmes/prstdbrnd1');
-        $grid = $crawler->filter(".grid");
-        $this->assertEquals(2, $grid->count());
+        $grid = $crawler->filter(".leading-gap")->filter(".grid");
+        $this->assertEquals(0, $grid->count());
```
Comment (votemike) [inline tests/Controller/FindByPid/Tlec/TlecControllerClipsListTest.php]: I believe there is an `assertCount` in PHPUnit that _might_ work here.

File: src/RecEng/RecEngService.php [hunk 7]
```diff
@@ -40,45 +42,54 @@ class RecEngService
...
+    ) {
         $this->client = $client;
         $this->audioKey = $audioKey;
         $this->videoKey = $videoKey;
         $this->programmesService = $programmesService;
         $this->router = $router;
         $this->logger = $logger;
         $this->cache = $cache;
+        $this->collapsedBroadcastsService = $collapsedBroadcastsService;
+        $this->aggregationService = $aggregationService;
```
Comment (tatitati) [inline src/RecEng/RecEngService.php]: I really think that this service is doing too much. Usually we have a service doing/fetching one thing. But this service is using more services. So this service has the responsibilities of: recEngService+CollapsedBroadcastService+aggregationsService.

I'm wondering if we can in this service just return the list of PIDs that we fetched from the HTTP request. 
And in the controller take these PIDs and invoke the rest of services. 

This simplify the design (more specialized) and is gonna be much easier to test. So far we have three different data sources that we have to stub in our mocks: from the HTTP requests, from the collapsedBroadcastRepository, and from AggregationRepository.

File: app/Resources/views/recommendations/dsamen_recommends_footer.html.twig [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                    'cta_options': {
+                        'cta_class': 'br-box-secondary',
+                        'show_duration': false,
+                    }
+                }) -}}
+            </li>
+        {%- endfor -%}
+    </ul>
+    <script>
+        require(['jquery-1.9','istats-1'], function($, istats) {
```
Comment (BjornTwachtmann) [inline app/Resources/views/recommendations/dsamen_recommends_footer.html.twig]: I'd probably omit this iStats code for now, and examine this when we come to refactor our iStats stuff. This is ugly as sin and I'd rather find a way to do this with data attrbitues (we might even already have one somewhere). 

Leave a note here saying todo instead?

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/DsAmen/_dsAmen-components.scss [hunk 9]
```diff
@@ -50,6 +50,7 @@ $ie: false !default;
 //@import 'Molecule/Cover/cover';
 @import 'Molecule/icon';
 //@import 'Molecule/Letterbox/letterbox';
+@import 'Molecule/lazyload';
```
Comment (BjornTwachtmann) [inline src/DsAmen/_dsAmen-components.scss]: won't work I think

**Pull request merged**